sympy/solvers/solvers.py

```diff
@@ -1557,7 +1557,11 @@ def _expand(p):
                         try:
                             t = Dummy('t')
                             iv = _solve(u - t, symbol, **flags)
-                            soln = list(ordered({i.subs(t, s) for i in iv for s in soln}))
+                            new_soln = list(ordered({i.subs(t, s) for i in iv for s in soln}))
+                            if new_soln[0] == S.NaN:
```

Comment:
> what to do with the rest?
 
yeah probably this would be a nicer approach that we remove all occurrences of `nan` and if there are solutions still in the list(`new_soln`) we pass it to `soln` and if the list is empty we leave the `soln`(making `soln` have only the roots calculated before). 

> Secondly, does a similar check need to be added for the case len(gens) > 1?

I guess so, there might be a case that would cause the same ambiguity. I am looking for such a case.

And regarding the performance issue, I'm not sure whether it would matter a great deal.